src/ring_theory/is_adjoin_root.lean

```diff
@@ -0,0 +1,578 @@
+/-
+Copyright (c) 2022 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen
+-/
+import data.polynomial.algebra_map
+import ring_theory.adjoin_root
+import ring_theory.power_basis
+
+/-!
+# A predicate on adjoining roots of polynomial
+
+This file defines a predicate `is_adjoin_root S f`, which states that the ring `S` can be
+constructed by adjoining a specified root of the polynomial `f : R[X]` to `R`.
+This predicate is useful when the same ring can be generated by adjoining the root of different
+polynomials, and you want to vary which polynomial you're considering.
+
+The results in this file are intended to mirror those in `ring_theory.adjoin_root`,
+in order to provide an easier way to translate results from one to the other.
+
+## Main definitions
+
+The two main predicates in this file are:
+
+ * `is_adjoin_root S f`: `S` is generated by adjoining a specified root of `f : R[X]` to `R`
+ * `is_adjoin_root_monic S f`: `S` is generated by adjoining a root of the monic polynomial
+   `f : R[X]` to `R`
+
+Using `is_adjoin_root` to map into `S`:
+
+ * `is_adjoin_root.map`: inclusion from `R[X]` to `S`
+ * `is_adjoin_root.root`: the specific root adjoined to `R` to give `S`
+
+Using `is_adjoin_root` to map out of `S`:
+
+ * `is_adjoin_root.repr`: choose a non-unique representative in `R[X]`
+ * `is_adjoin_root.lift`, `is_adjoin_root.lift_hom`: lift a morphism `R →+* T` to `S →+* T`
+ * `is_adjoin_root_monic.mod_by_monic_hom`: a unique representative in `R[X]` if `f` is monic
+
+## Main results
+
+ * `adjoin_root.is_adjoin_root` and `adjoin_root.is_adjoin_root_monic`:
+   `adjoin_root` satisfies the conditions on `is_adjoin_root`(`_monic`)
+ * `is_adjoin_root_monic.power_basis`: the `root` generates a power basis on `S` over `R`
+ * `is_adjoin_root.aequiv`: algebra isomorphism showing adjoining a root gives a unique ring
+   up to isomorphism
+ * `is_adjoin_root_eq.minpoly_eq`: the minimal polynomial of the adjoined root of `f` is equal to
+   `f`, if `f` is irreducible and monic, and `R` is a GCD domain
+-/
+
+open_locale polynomial
+open polynomial
+noncomputable theory
+
+universes u v
+
+section move_me
+
+end move_me
+
+/-- `is_adjoin_root S f` states that the ring `S` can be constructed by adjoining a specified root
+of the polynomial `f : R[X]` to `R`.
+
+Compare `power_basis R S`, which does not explicitly specify which polynomial we adjoin a root of
+(in particular `f` does not need to be the minimal polynomial of the root we adjoin),
+and `adjoin_root` which constructs a new type.
+
+This is not a typeclass because the choice of root given `S` and `f` is not unique.
+-/
+@[nolint has_nonempty_instance] -- This class doesn't really make sense on a predicate
+structure is_adjoin_root {R : Type u} (S : Type v) [comm_semiring R] [semiring S] [algebra R S]
```

Comment:
I have the impression that specifying this is exactly the same as specifying an equiv `R[X] / span {f} ≃ₐ[R] S`

Reply:
Indeed! I experimented with a few ways and found this the nicest phrasing to work with. One reason for not going with `adjoin_root R f ≃ₐ[R] S` (which is defeq to your formulation) is that we can now define the "predicate" independent of `adjoin_root`, and reducing explicit dependencies seems like a good move given mathlib's structure.